### PR TITLE
Add Community development fund balance

### LINKED ISSUE #554: [On-Chain-Governance] Move foundation address UTXO to community balance called "Community Development Fund"
By new hardfork creating new block should stop allocating coins (UTXO) to foundation address but generate tokens to new community balance

### PR DESCRIPTION
**Pull request by bvbfan**
/kind feature

#### What this PR does / why we need it:

Creates Community Development Funds balance and transfer community UTXO to tokens

Fixes #554

### CODE REVIEW

File: src/validation.cpp [hunk 1]
```diff
@@ -1977,7 +2010,11 @@ Res ApplyGeneralCoinbaseTx(CCustomCSView & mnview, CTransaction const & tx, int
     if (height >= consensus.AMKHeight)
     {
         CAmount foundationReward{0};
-        if (height >= consensus.EunosHeight)
+        if (height >= consensus.FortCanningHeight)
+        {
+            // no foundation reward check anymore
+        }
+        else if (height >= consensus.EunosHeight)
```
Comment (Bushstar) [inline src/validation.cpp:2017]: `if (height >= consensus.EunosHeight && height < consensus.FortCanningHeight)`

Comment (Bushstar) [reply src/validation.cpp:2017]: Maybe not, saw the code and that there's another else if after these two!

**Comment by Bushstar** [comment]
My suggestion was a two step approach, first hard fork would switch foundation UTXO reward to community fund account layer reward, after which we move the funds to the account layer on the same address. Second hard fork we move the funds from the account layer to the community balance.

I personally do not like trying to null things on the UTXO layer, in this case it will look like the funds are still there on explorers even though that are unspendable on the consensus layer and now there's a balance on the account layer. If we do a two step and move funds to the account layer first we can make sure that the account layer balance is properly destroyed.

File: src/validation.cpp [hunk 2]
```diff
@@ -1414,7 +1414,10 @@ bool CheckBurnSpend(const CTransaction &tx, const CCoinsViewCache &inputs)
     Coin coin;
     for(size_t i = 0; i < tx.vin.size(); ++i)
     {
-        if (inputs.GetCoin(tx.vin[i].prevout, coin) && coin.out.scriptPubKey == Params().GetConsensus().burnAddress)
+        if (inputs.GetCoin(tx.vin[i].prevout, coin)
+        && (coin.out.scriptPubKey == Params().GetConsensus().burnAddress
+        || (coin.nHeight > Params().GetConsensus().FortCanningHeight
+        && coin.out.scriptPubKey == Params().GetConsensus().foundationShareScript)))
```
Comment (Bushstar) [inline src/validation.cpp:1420]: I would not treat foundationShareScript as burn, we will manually send whatever is left to the burn address.

File: src/validation.cpp [hunk 3]
```diff
@@ -1748,9 +1751,11 @@ DisconnectResult CChainState::DisconnectBlock(const CBlock& block, const CBlockI
                 }
 
                 // Check for burn outputs
-                if (tx.vout[o].scriptPubKey == Params().GetConsensus().burnAddress)
+                if (tx.vout[o].scriptPubKey == Params().GetConsensus().burnAddress
+                || (pindex->nHeight > Params().GetConsensus().FortCanningHeight
+                && tx.vout[o].scriptPubKey == Params().GetConsensus().foundationShareScript))
                 {
-                    eraseBurnEntries.push_back({Params().GetConsensus().burnAddress, static_cast<uint32_t>(pindex->nHeight), static_cast<uint32_t>(i)});
+                    eraseBurnEntries.push_back({tx.vout[o].scriptPubKey, static_cast<uint32_t>(pindex->nHeight), static_cast<uint32_t>(i)});
```
Comment (Bushstar) [inline src/validation.cpp:1758]: Same here, I'd not treat as burn.

File: src/validation.cpp [hunk 4]
```diff
@@ -2502,9 +2527,11 @@ bool CChainState::ConnectBlock(const CBlock& block, CValidationState& state, CBl
         // Search for burn outputs
         for (uint32_t j = 0; j < tx.vout.size(); ++j)
         {
-            if (tx.vout[j].scriptPubKey == Params().GetConsensus().burnAddress)
+            if (tx.vout[j].scriptPubKey == Params().GetConsensus().burnAddress
+            || (pindex->nHeight > Params().GetConsensus().FortCanningHeight
+            && tx.vout[j].scriptPubKey == Params().GetConsensus().foundationShareScript))
             {
-                writeBurnEntries.push_back({{Params().GetConsensus().burnAddress, static_cast<uint32_t>(pindex->nHeight), i},
+                writeBurnEntries.push_back({{tx.vout[j].scriptPubKey, static_cast<uint32_t>(pindex->nHeight), i},
```
Comment (Bushstar) [inline src/validation.cpp:2534]: Same here.

File: test/functional/feature_community_development_funds.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        history = node1.listburnhistory()
+        assert_equal(len(history), 1)
+        assert_equal(history[0]['owner'], '2NCWAKfEehP3qibkLKYQjXaWMK23k4EDMVS')
+        assert_equal(history[0]['blockHeight'], 202)
+        assert_equal(history[0]['txid'], txid)
+        assert_equal(history[0]['amounts'], ['5.00000000@DFI'])
+        # foundation coins received after fortcanningheight are burnt
+        foundationBalance = node1.getbalances()['mine']['trusted'] - Decimal("0.0010000")
+        assert_raises_rpc_error(-26, "burnt-output", node1.utxostoaccount, {'2NCWAKfEehP3qibkLKYQjXaWMK23k4EDMVS': str(foundationBalance) + "@0"})
+
```
Comment (Bushstar) [inline test/functional/feature_community_development_funds.py:52]: Please add a test for rollback over hard fork, do the balance changes from account to community balance revert?

Comment (bvbfan) [reply test/functional/feature_community_development_funds.py:62]: Added rollback test

**Review by Bushstar — APPROVED** [review]

**Pull request merged**